File: bin/partition_multiqc.rb
#!/usr/bin/env ruby

# == USAGE
# ./partition_multiqc.rb [ -h | --help ]
#[ -i | --infile ] |[ -o | --outfile ] |
# == DESCRIPTION
# Run MultiQC on subset of files based on common grouping variable
#
# == OPTIONS
# -h,--help Show help
# -i,--infile=INFILE input file
# -o,--outfile=OUTFILE : output file

#
# == EXPERT OPTIONS
#
# == AUTHOR
#  Marc Hoeppner, mphoeppner@gmail.com

require 'optparse'
require 'ostruct'

### Define modules and classes here

def run(command)

	warn "#{Time.now}\tExecuting: #{command}"
	system(command)
end

### Get the script arguments and open relevant files
options = OpenStruct.new()
opts = OptionParser.new()
opts.banner = "A script description here"
opts.separator ""
opts.on("-c","--chunk", "=CHUNK","Chunk size") {|argument| options.chunk = argument }
opts.on("-b","--title", "=TITLE","Report title") {|argument| options.title = argument }
opts.on("-k","--config", "=CONFIG","QC config") {|argument| options.config = argument }
opts.on("-n","--name", "=NAME","Name of project") {|argument| options.name = argument }
opts.on("-h","--help","Display the usage information") {
 puts opts
 exit
}

opts.parse!

files = Dir["*_001*"]

grouped = files.group_by {|f| f.split(/_[RI][0-9]_/)[0] }

options.chunk ? chunk = options.chunk.to_i : chunk = 100

grouped.each_slice(chunk) do |slice|

	groups = []

        list = File.new("list.txt","w+")
	slice.each do |g,reports|
		groups << g
		reports.each {|r| list.puts r }		
	end
	list.close

	first_lib = groups.sort[0]
	last_lib = groups.sort[-1]

	command = "multiqc -n multiqc_report_#{options.name}_#{first_lib}-#{last_lib}.html -b \"#{options.title}\" -c #{options.config} --file-list list.txt"
	run(command)

end

File: bin/partition_multiqc.pl
#!/usr/bin/env perl

use strict;
use Getopt::Long;

my $usage = qq{
perl my_script.pl
  Getting help:
    [--help]

  Input:
    [--chunk int]
		Size of each chunk

    [--title STR]
		Title of the report

    [--config FILE]
		Name of the config 

    [--name STR]
		LIMS Project name

};

my $outfile = undef;
my $chunk = 100;
my $title = undef;
my $name = undef;
my $config = undef;

my $help;

GetOptions(
    "help" => \$help,
    "chunk=i" => \$chunk,
    "title=s" => \$title,
    "config=s" => \$config,
    "name=s" => \$name);

# Print Help and exit
if ($help) {
    print $usage;
    exit(0);
}


my @files = glob("*.tsv *.zip *_mqc.out") ;
my %bucket;

die "Must provide a name (--name) and title (--title)!\n" unless (defined $name && defined $title);
my @general;

foreach my $file (@files) {

	my $trunk = "";

	if ($file =~ /.*_mqc.out/) {
		push(@general,$file);
	} elsif ($file =~ /.*_summary.tsv/) {
		$trunk = (split /_summary/, $file)[0];
	} elsif ($file =~ /.*_[RI][0-9]_001.*/) {
		$trunk = (split /_[RI][0-9]_001/, $file)[0] ;
	} else {
		$trunk = (split /\./, $file)[0];
	}	
	next if ($trunk eq "");

	if (exists $bucket{$trunk}) {
		push( @{ $bucket{$trunk} }, $file )
	} else {
		$bucket{$trunk} = [ $file ];
	}

}

my @output;
my $counter = 0;
my $first_key = undef;
my $this_key = undef;

foreach my $key (sort keys %bucket ) {

	unless (defined $first_key) {
		$first_key = $key ;
	}

	my @entries = @{ $bucket{$key} } ;

	if ($counter > $chunk) {
	
		foreach my $g (@general) {
                	push(@output,$g);
        	}
		printf STDERR "Counter is: " .  $counter . "\n";
		my $file_name = "multiqc_report_" . $name . "_" . $first_key . "-" . $this_key . ".html" ;
		
		run_multiqc($file_name,$title,$config,@output);

		@output = ();
		$counter = 0;
		$first_key = $key ;

	}

	$counter += 1;

	foreach my $e (@entries) {
		push(@output,$e);
	}

	$this_key = $key;

}

my $file_name = "multiqc_report_" . $name . "_" . $first_key . "-" . $this_key . ".html";
run_multiqc($file_name,$title,$config,@output);

sub run_multiqc {

	my($file_name,$title,$config,@entries) = @_ ;

	open(FILE,'>', "files.txt");

	foreach my $e (@entries) {
		printf FILE $e . "\n";
	}

	close(FILE);

	my $command = "multiqc -n $file_name -b \"" . $title . "\" -c $config --file-list files.txt" ;
	system($command);

}	

